Modernize LoggingTensorMode

soulitzer opened the pull request:
This PR:
- updates LoggingTensorMode to use the new PythonDispatchMode
- updates some tests that use LoggingTensorMode to properly use the new version

Open questions:
- why is the old python mode still working? Do we intend to keep it? (a small bit of additional logic is necessary to support it)
  - This PR updates some of the tests that test old python mode behavior (e.g., all outputs are wrapped). If we intend to support old python mode long term we should bring it back, and just add new tests instead.

TODO (this PR):
- Test nesting behavior

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77667
* #77743

See https://github.com/pytorch/pytorch/issues/77544

soulitzer committed:
Modernize LoggingTensorMode

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Modernize LoggingTensorMode"


See https://github.com/pytorch/pytorch/issues/77544

[ghstack-poisoned]

soulitzer commented on torch/utils/_mode_utils.py:
```diff
@@ -71,7 +71,7 @@ def _enable_mode(mode, mode_info: _ModeInfo, *, replace=None, ignore_preexisting
                          f'or None as an argument got {type(mode)} instead')
     old = mode_info.get_mode()
-    if old is mode:
+    if type(old) is type(mode):
```
This may no longer support the old version of python mode

zou3519 replied:
Just checking, this still needs to be fixed because we can have multiple instances of the same class, right? (When this does get fix we should add a test for it)

soulitzer committed:
Update on "Modernize LoggingTensorMode"

This PR:
- updates LoggingTensorMode to use the new PythonDispatchMode
- updates some tests that use LoggingTensorMode to properly use the new version

Open questions:
- why is the old python mode still working? Do we intend to keep it? (a small bit of additional logic is necessary to support it)


See https://github.com/pytorch/pytorch/issues/77544

[ghstack-poisoned]

ezyang commented:
IMO we should just dump old python mode

soulitzer commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -114,14 +120,19 @@ def log_input(name: str, var: object):
+
+@contextlib.contextmanager
+def capture_logs_with_logging_tensor_mode():
+    with push_torch_dispatch_mode(LoggingTensorMode(inner=None)), capture_logs(True) as logs:
```
TODO: shouldn't don't need inner

ezyang replied:
@samdow yeah this looks like we should just support direct constructor lol

samdow replied:
We should also add a test for this function

I'm pretty sure that this will error right now for the inner reason. Also flags to me that I really dislike that the argument for `push_torch_dispatch_mode` is going to be a constructor for a Mode while `enable_torch_dispatch_mode` is going to be the Mode itself. TODO (@samdow): fix that

soulitzer committed:
Update on "Modernize LoggingTensorMode"

This PR:
- updates LoggingTensorMode to use the new PythonDispatchMode
- updates some tests that use LoggingTensorMode to properly use the new version

Open questions:
- why is the old python mode still working? Do we intend to keep it? (a small bit of additional logic is necessary to support it)
  - This PR updates some of the tests that test old python mode behavior (e.g., all outputs are wrapped). If we intend to support old python mode long term we should bring it back, and just add new tests instead.

TODO (this PR):
- Test nesting behavior


See https://github.com/pytorch/pytorch/issues/77544

[ghstack-poisoned]

soulitzer committed:
Update on "Modernize LoggingTensorMode"

This PR:
- updates LoggingTensorMode to use the new PythonDispatchMode
- updates some tests that use LoggingTensorMode to properly use the new version

Open questions:
- why is the old python mode still working? Do we intend to keep it? (a small bit of additional logic is necessary to support it)
  - This PR updates some of the tests that test old python mode behavior (e.g., all outputs are wrapped). If we intend to support old python mode long term we should bring it back, and just add new tests instead.

TODO (this PR):
- Test nesting behavior


See https://github.com/pytorch/pytorch/issues/77544

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -114,14 +120,19 @@ def log_input(name: str, var: object):
 
 @contextlib.contextmanager
-def capture_logs() -> Iterator[List[str]]:
+def capture_logs(is_dispatch_mode=False) -> Iterator[List[str]]:
```
To check my understanding, `is_dispatch_mode` switches on if we are using Logging Tensor, or if we are using the Mode, right?

If so, then nit: the name is a bit ambiguous, if `is_dispatch_mode=False` then sometimes "torch_function mode" comes into mind. Would just add a comment here for people coming in the future, or change this to `is_mode`

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -1001,9 +1001,10 @@ def wrap(e):
     def test_storage_can_be_converted_to_python_object(self):
-        with enable_torch_dispatch_mode(LoggingTensorMode):
+        # Is this test still relevant? We're just dealing with plain tensors now.
+        with enable_torch_dispatch_mode(LoggingTensor):
             s = torch.Storage()
-            z = LoggingTensorMode(torch.empty([]))
+            z = LoggingTensor(torch.empty([]))
```
I think this is here to as a test case for a bug reported by @cbalioglu.

IIRC what we're testing here is that Storage actually works with torch_dispatch objects. So I don't think we need the mode part of it (enable_torch_dispatch_mode) and should just check that LoggingTensor(...).set_(storage) works

soulitzer committed:
Update on "Modernize LoggingTensorMode"

This PR:
- updates LoggingTensorMode to use the new PythonDispatchMode
- updates some tests that use LoggingTensorMode to properly use the new version

Open questions:
- why is the old python mode still working? Do we intend to keep it? (a small bit of additional logic is necessary to support it)
  - This PR updates some of the tests that test old python mode behavior (e.g., all outputs are wrapped). If we intend to support old python mode long term we should bring it back, and just add new tests instead.

TODO (this PR):
- Test nesting behavior


See https://github.com/pytorch/pytorch/issues/77544

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge this on green

soulitzer commented:
@pytorchbot merge this

soulitzer commented:
@pytorchbot merge this
